File: src/orchestrator/board-recovery.ts
import { accessSync, constants, existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs';
import { spawnSync } from 'node:child_process';
import { join } from 'node:path';
import { BOARD_FILE } from './types.js';

/**
 * Orchestrator board recovery (PRD:888 Q19) — the board-recovery decisions
 * folded out of scripts/orchestrate-loop.sh into typed, tested code. The
 * shell is a thin caller via `devagent board-recovery` (src/cli.ts), using
 * the same seam pattern as selfbuild-gate.ts (loops 121/122): the gate
 * PERFORMS the action it verdicts (requeue write, board archive, merged-
 * worktree prune) and prints exactly one verdict line
 * `<wait|requeue|archive>: <reason>`; the shell only maps the verdict to
 * loop control:
 *
 *   wait     sleep POLL_SECS, continue (parked below threshold, requeue
 *            disabled, or a completed board already archived by the gate)
 *   requeue  parked tasks reset to pending; the shell clears its poll
 *            counter, sleeps, and lets the next cycle dispatch them
 *   archive  stuck board archived; the shell falls through to the queue
 *            bridge THIS cycle (#73: previously the parked block always did
 *            `sleep POLL_SECS; continue`, leaving the board absent and the
 *            factory idle for a whole poll interval)
 *
 * Exit-code contract is backlog-check's (PRD:889): 0 = verdict printed,
 * 2 = unresolved (caller misuse / board vanished mid-cycle). A crashed CLI
 * also exits nonzero, so the shell only honors a verdict word it can read
 * back from the output — the fallback is wait: a crashed gate must never
 * archive a board or hot-loop the factory.
 *
 * Decision table (ported verbatim from the shell, scripts/orchestrate-loop.sh
 * :138-196): the gate is consulted only when the board has no open tasks
 * (open = not done/failed/blocked) and at least one task.
 *   - every task done           → archive the completed board so the next
 *                                 iteration re-bridges scouted queue items
 *                                 and plans a fresh board from the goal
 *                                 (infinity cycle, #42); merged worktrees
 *                                 are pruned alongside. Verdict: wait.
 *   - all failed/blocked,
 *     requeue disabled          → wait (parked, logged)
 *   - all failed/blocked,
 *     parked polls < threshold  → wait (parked, logged with P/Q counter)
 *   - threshold reached         → requeue (failed/blocked → pending,
 *                                 attempts reset to 0), then re-inspect:
 *     still failed/blocked      → archive board-stuck (defensive: requeue
 *                                 clears every dispatch-dead status, so this
 *                                 only fires if the reset write failed)
 *     every task pending        → the scheduler still cannot dispatch
 *                                 (readiness never recomputes for a board
 *                                 with no done tasks to unblock deps), so
 *                                 archive board-stuck and let the queue
 *                                 bridge take over
 *     otherwise                 → requeue verdict (some tasks done; the
 *                                 reset pending tasks dispatch next cycle)
 *
 * The parked-poll counter is LOOP state, not board state: the shell owns it
 * (incremented per parked cycle, cleared on any action verdict or when the
 * board has open work) and passes the post-increment value via
 * --parked-polls. Folding it into the gate would require persisting driver
 * state the board file does not carry.
 */

/** Statuses the scheduler can never dispatch again (the dispatch-dead pair). */
const DEAD_STATUSES: readonly string[] = ['failed', 'blocked'];

/** Board task shape as read back from JSON (statuses are validated loosely, mirroring the shell's `node -e` leniency). */
interface BoardTaskLike {
  status?: unknown;
  attempts?: unknown;
}

/** Parsed board file: the raw object (written back verbatim after a requeue)
 * plus its validated tasks array — the same array reference `raw.tasks` holds. */
interface BoardFile {
  raw: object;
  tasks: BoardTaskLike[];
}

/** Raw task counts the decision table reads, mirroring the shell's board_* helpers. */
export interface BoardCounts {
  total: number;
  done: number;
  /** Not done/failed/blocked — everything the loop may still dispatch. */
  open: number;
  /** failed/blocked — the dispatch-dead states. */
  stuck: number;
  pending: number;
}

export interface RecoveryOptions {
  /** Parked cycles including this one (the shell's post-increment counter). */
  parkedPolls: number;
  /** Requeue threshold in parked polls; 0 = never requeue. */
  requeueAfter: number;
  /** Loop sleep seconds, quoted in wait/requeue verdict text. */
  pollSecs: number;
}

export type RecoveryAction = 'wait' | 'requeue' | 'archive';

export interface BoardRecoveryVerdict {
  action: RecoveryAction;
  reason: string;
}

/** First-cycle intent: what the parked board needs before any mutation. */
export type RecoveryIntent =
  | { kind: 'archive-complete' }
  | { kind: 'requeue-now' }
  | { kind: 'wait'; reason: string };

/** Post-requeue intent: whether the reset board is still undispatchable. */
export type PostRequeueIntent =
  | { kind: 'archive'; detail: string }
  | { kind: 'requeue' };

/** Count the board the way the shell's board_* helpers did. */
export function countBoard(tasks: readonly BoardTaskLike[]): BoardCounts {
  let done = 0;
  let stuck = 0;
  let pending = 0;
  for (const t of tasks) {
    const status = t?.status;
    if (status === 'done') done++;
    else if (DEAD_STATUSES.includes(String(status))) stuck++;
    if (status === 'pending') pending++;
  }
  const total = tasks.length;
  return { total, done, open: total - done - stuck, stuck, pending };
}

/**
 * Pre-mutation decision: completed board, parked wait, or threshold reached.
 * `open > 0` / `total === 0` are defensive (the shell guard keeps the gate
 * out of those paths); they verdict wait, never an action.
 */
export function decideBoardRecovery(counts: BoardCounts, opts: RecoveryOptions): RecoveryIntent {
  if (counts.total === 0) return { kind: 'wait', reason: 'board has no tasks' };
  if (counts.open > 0) return { kind: 'wait', reason: `board has ${counts.open} open task(s)` };
  if (counts.done === counts.total) return { kind: 'archive-complete' };
  const sleep = `; sleeping ${opts.pollSecs}s`;
  if (opts.requeueAfter <= 0) {
    return { kind: 'wait', reason: `${counts.stuck} task(s) failed/blocked${sleep} (requeue disabled)` };
  }
  if (opts.parkedPolls < opts.requeueAfter) {
    return {
      kind: 'wait',
      reason: `${counts.stuck} task(s) failed/blocked (${opts.parkedPolls}/${opts.requeueAfter})${sleep}`,
    };
  }
  return { kind: 'requeue-now' };
}

/**
 * Post-requeue decision, over the counts recomputed AFTER the reset write.
 * The stuck branch is defensive table parity — requeue clears every
 * dispatch-dead status (and a failed reset write throws, exiting the gate
 * unresolved), so it cannot fire in practice. The all-pending branch is the
 * real one: a board with no done task never recomputes readiness, so
 * requeue cannot unstick it; archive and let the queue bridge take over.
 */
export function decidePostRequeue(counts: BoardCounts): PostRequeueIntent {
  if (counts.stuck > 0) return { kind: 'archive', detail: `board stuck (${counts.stuck} failed/blocked)` };
  if (counts.total > 0 && counts.pending === counts.total) {
    return { kind: 'archive', detail: 'board all-pending but undispatchable' };
  }
  return { kind: 'requeue' };
}

/** Render a verdict the way the CLI prints it: the shell parses the word before the first colon. */
export function formatVerdict(verdict: BoardRecoveryVerdict): string {
  return `${verdict.action}: ${verdict.reason}`;
}

/** `date +%Y%m%d-%H%M%S` in local time — the archive filename stamp the shell used. */
export function formatTimestamp(d: Date): string {
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${d.getFullYear()}${pad(d.getMonth() + 1)}${pad(d.getDate())}-${pad(d.getHours())}${pad(d.getMinutes())}${pad(d.getSeconds())}`;
}

function readBoard(boardPath: string): BoardFile | null {
  let parsed: unknown;
  try {
    parsed = JSON.parse(readFileSync(boardPath, 'utf8'));
  } catch {
    return null;
  }
  if (!parsed || typeof parsed !== 'object' || !('tasks' in parsed)) return null;
  if (!Array.isArray(parsed.tasks)) return null;
  // Elements stay unknown-shaped: every access below is optional-guarded,
  // mirroring the shell's `t.status` leniency on hand-edited boards.
  const tasks = parsed.tasks as BoardTaskLike[];
  return { raw: parsed, tasks };
}

/**
 * Port of requeue_parked(): failed/blocked → pending with the attempts
 * budget reset; writes only when something changed, preserving the shell's
 * byte-for-byte JSON shape (2-space indent, trailing newline). tmp+rename
 * so a crash mid-write cannot corrupt the board.
 */
export function requeueParked(boardPath: string, board: BoardFile): number {
  let n = 0;
  for (const t of board.tasks) {
    if (t && DEAD_STATUSES.includes(String(t.status))) {
      t.status = 'pending';
      t.attempts = 0;
      n++;
    }
  }
  if (n > 0) {
    const tmp = `${boardPath}.tmp`;
    writeFileSync(tmp, JSON.stringify(board.raw, null, 2) + '\n');
    renameSync(tmp, boardPath);
  }
  return n;
}

/** Move the board into .devagent/archive/ under `prefix-<stamp>.json`; returns the repo-relative path. */
export function archiveBoard(
  repoPath: string,
  boardPath: string,
  prefix: 'board' | 'board-stuck',
  stamp: string,
): string {
  const dir = join(repoPath, '.devagent', 'archive');
  mkdirSync(dir, { recursive: true });
  const rel = join('.devagent', 'archive', `${prefix}-${stamp}.json`);
  renameSync(boardPath, join(repoPath, rel));
  return rel;
}

/** Safe-gated merged-worktree prune (the shell's cleanup_merged_worktrees): only an executable repo script, never fatal. */
function pruneMergedWorktrees(repoPath: string): void {
  const script = join(repoPath, 'scripts', 'git-cleanup-merged.sh');
  if (!existsSync(script)) return;
  try {
    accessSync(script, constants.X_OK);
  } catch {
    return;
  }
  try {
    spawnSync(script, ['--root', repoPath, '--apply'], { stdio: 'ignore' });
  } catch {
    // cleanup is opportunistic; a failed prune must not break the cycle
  }
}

export interface RunBoardRecoveryOptions extends RecoveryOptions {
  /** Injectable clock for the archive stamp (tests). */
  now?: Date;
}

/**
 * One recovery cycle: read the board, decide, perform the action, return the
 * verdict. Mutates the board file (requeue) and moves it (archive) exactly
 * where the shell did. Throws only when the board vanishes between the
 * decision and the action — the CLI maps that to exit 2 (unresolved → the
 * shell falls back to wait).
 */
export function runBoardRecovery(repoPath: string, opts: RunBoardRecoveryOptions): BoardRecoveryVerdict {
  const boardPath = join(repoPath, BOARD_FILE);
  const board = readBoard(boardPath);
  if (!board) return { action: 'wait', reason: 'board unreadable' };
  const stamp = formatTimestamp(opts.now ?? new Date());
  const sleep = `; sleeping ${opts.pollSecs}s`;

  const counts = countBoard(board.tasks);
  const intent = decideBoardRecovery(counts, opts);
  if (intent.kind === 'wait') return { action: 'wait', reason: intent.reason };
  if (intent.kind === 'archive-complete') {
    const rel = archiveBoard(repoPath, boardPath, 'board', stamp);
    pruneMergedWorktrees(repoPath);
    return { action: 'wait', reason: `board complete (${counts.done} done); archived to ${rel}` };
  }

  const reset = requeueParked(boardPath, board);
  const post = decidePostRequeue(countBoard(board.tasks));
  if (post.kind === 'requeue') {
    return { action: 'requeue', reason: `reset ${reset} parked task(s) to pending${sleep}` };
  }
  const rel = archiveBoard(repoPath, boardPath, 'board-stuck', stamp);
  return {
    action: 'archive',
    reason: `reset ${reset} parked task(s) to pending; ${post.detail}; archived to ${rel}`,
  };
}

File: internal/orchestrator/board-recovery.go
// Package file mirrors src/orchestrator/board-recovery.ts (FR-GO-07, issue #194).
package orchestrator

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/FreePeak/devagent/internal/config"
)

// deadStatuses mirrors the TS DEAD_STATUSES: statuses the scheduler can
// never dispatch again (the dispatch-dead pair).
var deadStatuses = []string{"failed", "blocked"}

// BoardTaskLike mirrors the TS BoardTaskLike: board task shape as read back
// from JSON (statuses validated loosely, mirroring the shell's `node -e`
// leniency).
type BoardTaskLike struct {
	Status        any `json:"status"`
	Attempts      any `json:"attempts"`
	TotalAttempts any `json:"totalAttempts"`
}

// boardFile mirrors the TS BoardFile: the raw parsed object (written back
// verbatim after a requeue) plus its validated tasks array — the same maps
// raw["tasks"] holds, so mutations surface on the re-marshal.
//
// Deviation vs the TS: JS JSON.stringify preserves key insertion order; Go's
// map marshal sorts keys. Requeue write-backs keep every field and value but
// reorder object keys alphabetically.
type boardFile struct {
	Raw   map[string]any
	Tasks []map[string]any
}

// BoardCounts mirrors the TS BoardCounts: raw task counts the decision
// table reads.
type BoardCounts struct {
	Total   int `json:"total"`
	Done    int `json:"done"`
	Open    int `json:"open"`  // not done/failed/blocked — everything the loop may still dispatch
	Stuck   int `json:"stuck"` // failed/blocked — the dispatch-dead states
	Pending int `json:"pending"`
}

// RecoveryOptions mirrors the TS RecoveryOptions.
type RecoveryOptions struct {
	// ParkedPolls: parked cycles including this one (the shell's
	// post-increment counter).
	ParkedPolls int
	// RequeueAfter: requeue threshold in parked polls; 0 = never requeue.
	RequeueAfter int
	// PollSecs: loop sleep seconds, quoted in wait/requeue verdict text.
	PollSecs int
	// MaxTotalAttempts: cumulative lifetime dispatch cap (Q17/Q36); a
	// failed/blocked task whose totalAttempts reaches it is refused the
	// requeue reset and stays terminal. 0 = unbounded.
	MaxTotalAttempts int
}

// RecoveryAction mirrors the TS 'wait' | 'requeue' | 'archive' union.
type RecoveryAction string

const (
	RecoveryActionWait    RecoveryAction = "wait"
	RecoveryActionRequeue RecoveryAction = "requeue"
	RecoveryActionArchive RecoveryAction = "archive"
)

// BoardRecoveryVerdict mirrors the TS BoardRecoveryVerdict.
type BoardRecoveryVerdict struct {
	Action RecoveryAction `json:"action"`
	Reason string         `json:"reason"`
}

// Recovery intent kinds, mirroring the TS RecoveryIntent discriminated union.
const (
	IntentArchiveComplete = "archive-complete"
	IntentRequeueNow      = "requeue-now"
	IntentWait            = "wait"
)

// PostRequeueIntentKindArchive / ...KindRequeue mirror the TS
// PostRequeueIntent discriminated union.
const (
	PostRequeueIntentKindArchive = "archive"
	PostRequeueIntentKindRequeue = "requeue"
)

// RecoveryIntent mirrors the TS RecoveryIntent union as a tagged struct
// (Go has no discriminated unions): Kind is 'archive-complete' (archive the
// completed board), 'requeue-now' (threshold reached), or 'wait' with a
// Reason.
type RecoveryIntent struct {
	Kind   string
	Reason string // wait only
}

// PostRequeueIntent mirrors the TS PostRequeueIntent union as a tagged
// struct: Kind is 'archive' with a human-readable Detail, or 'requeue'.
type PostRequeueIntent struct {
	Kind   string
	Detail string // archive only
}

// CountBoard mirrors countBoard: count the board the way the shell's
// board_* helpers did. Malformed rows (missing/numeric status) count toward
// total and open.
func CountBoard(tasks []BoardTaskLike) BoardCounts {
	done := 0
	stuck := 0
	pending := 0
	for _, t := range tasks {
		status := looseString(t.Status)
		if status == "done" {
			done++
		} else if containsString(deadStatuses, status) {
			stuck++
		}
		if status == "pending" {
			pending++
		}
	}
	total := len(tasks)
	return BoardCounts{Total: total, Done: done, Open: total - done - stuck, Stuck: stuck, Pending: pending}
}

// DecideBoardRecovery mirrors decideBoardRecovery: pre-mutation decision —
// completed board, parked wait, or threshold reached. open > 0 / total == 0
// are defensive (the shell guard keeps the gate out of those paths); they
// verdict wait, never an action.
func DecideBoardRecovery(counts BoardCounts, opts RecoveryOptions) RecoveryIntent {
	if counts.Total == 0 {
		return RecoveryIntent{Kind: IntentWait, Reason: "board has no tasks"}
	}
	if counts.Open > 0 {
		return RecoveryIntent{Kind: IntentWait, Reason: fmt.Sprintf("board has %d open task(s)", counts.Open)}
	}
	if counts.Done == counts.Total {
		return RecoveryIntent{Kind: IntentArchiveComplete}
	}
	sleep := fmt.Sprintf("; sleeping %ds", opts.PollSecs)
	if opts.RequeueAfter <= 0 {
		return RecoveryIntent{Kind: IntentWait, Reason: fmt.Sprintf("%d task(s) failed/blocked%s (requeue disabled)", counts.Stuck, sleep)}
	}
	if opts.ParkedPolls < opts.RequeueAfter {
		return RecoveryIntent{Kind: IntentWait, Reason: fmt.Sprintf("%d task(s) failed/blocked (%d/%d)%s", counts.Stuck, opts.ParkedPolls, opts.RequeueAfter, sleep)}
	}
	return RecoveryIntent{Kind: IntentRequeueNow}
}

// DecidePostRequeue mirrors decidePostRequeue: post-requeue decision, over
// the counts recomputed AFTER the reset write. The stuck branch fires when
// the cumulative attempt cap (Q17/Q36) refused the reset for at least one
// dead task; the all-pending branch is the original real one (a board with
// no done task never recomputes readiness, so requeue cannot unstick it).
func DecidePostRequeue(counts BoardCounts) PostRequeueIntent {
	if counts.Stuck > 0 {
		return PostRequeueIntent{Kind: PostRequeueIntentKindArchive, Detail: fmt.Sprintf("board stuck (%d failed/blocked)", counts.Stuck)}
	}
	if counts.Total > 0 && counts.Pending == counts.Total {
		return PostRequeueIntent{Kind: PostRequeueIntentKindArchive, Detail: "board all-pending but undispatchable"}
	}
	return PostRequeueIntent{Kind: PostRequeueIntentKindRequeue}
}

// FormatVerdict mirrors formatVerdict: render a verdict the way the CLI
// prints it — the shell parses the word before the first colon.
func FormatVerdict(verdict BoardRecoveryVerdict) string {
	return fmt.Sprintf("%s: %s", verdict.Action, verdict.Reason)
}

// FormatTimestamp mirrors formatTimestamp: `date +%Y%m%d-%H%M%S` in the
// time's own location — the archive filename stamp the shell used.
func FormatTimestamp(d time.Time) string {
	return d.Format("20060102-150405")
}

// readBoard mirrors the TS readBoard: parse the raw board file, requiring an
// object with an array `tasks`. Elements stay unknown-shaped (possibly nil
// maps): every access is guarded, mirroring the TS `t.status` leniency on
// hand-edited boards.
func readBoard(boardPath string) *boardFile {
	data, err := os.ReadFile(boardPath)
	if err != nil {
		return nil
	}
	var parsed map[string]any
	if json.Unmarshal(data, &parsed) != nil {
		return nil
	}
	rawTasks, ok := parsed["tasks"].([]any)
	if !ok {
		return nil
	}
	tasks := make([]map[string]any, len(rawTasks))
	for i, el := range rawTasks {
		if m, ok := el.(map[string]any); ok {
			tasks[i] = m
		}
	}
	return &boardFile{Raw: parsed, Tasks: tasks}
}

// RequeueResult mirrors the TS RequeueResult: resets performed vs refused by
// the cumulative cap.
type RequeueResult struct {
	// Reset: dead tasks reset to pending with a fresh per-round attempts budget.
	Reset int `json:"reset"`
	// Capped: dead tasks at/above the cumulative cap — refused, left terminal (Q17/Q36).
	Capped int `json:"capped"`
}

// RequeueParked mirrors requeueParked (the port of requeue_parked()):
// failed/blocked → pending with the attempts budget reset; writes only when
// something changed (2-space indent, trailing newline, tmp+rename so a crash
// mid-write cannot corrupt the board). Above the cumulative cap (Q17/Q36,
// 0 = unbounded) a dead task is refused the fresh budget and stays
// failed/blocked; totalAttempts itself is never reset. Missing or
// non-numeric lifetime history counts as zero.
func RequeueParked(boardPath string, board *boardFile, maxTotalAttempts int) RequeueResult {
	reset := 0
	capped := 0
	for _, t := range board.Tasks {
		if t == nil || !containsString(deadStatuses, looseString(t["status"])) {
			continue
		}
		total, finite := numberIsFinite(t["totalAttempts"])
		if maxTotalAttempts > 0 && finite && total >= float64(maxTotalAttempts) {
			capped++
			continue
		}
		t["status"] = "pending"
		t["attempts"] = 0
		reset++
	}
	if reset > 0 {
		writeJSONAtomic(boardPath, board.Raw)
	}
	return RequeueResult{Reset: reset, Capped: capped}
}

// ArchiveRetentionKeep mirrors the TS ARCHIVE_RETENTION_KEEP: default bound
// for .devagent/archive/ retention (Q16) — the newest N stamped archives are
// kept.
const ArchiveRetentionKeep = 20

// archiveName mirrors the TS ARCHIVE_NAME regexp: archive filenames carry a
// `YYYYMMDD-HHMMSS` stamp; only these are pruned.
var archiveName = regexp.MustCompile(`^(?:board|board-stuck)-(\d{8}-\d{6})\.json$`)

// PruneArchive mirrors pruneArchive: bounded retention prune of
// .devagent/archive/ (Q16) — keep the newest `keep` stamped archives, delete
// the rest, ordered by the filename stamp (deterministic under an injected
// clock). keep <= 0 is unbounded (never prune). Best-effort: a failed unlink
// is swallowed. Returns the removed names.
func PruneArchive(repoPath string, keep int) []string {
	if keep <= 0 {
		return []string{}
	}
	dir := filepath.Join(repoPath, ".devagent", "archive")
	entries, err := os.ReadDir(dir)
	if err != nil {
		return []string{} // no archive dir yet — nothing to prune
	}
	type stamped struct {
		name  string
		stamp string
	}
	var stamps []stamped
	for _, e := range entries {
		if m := archiveName.FindStringSubmatch(e.Name()); m != nil {
			stamps = append(stamps, stamped{name: e.Name(), stamp: m[1]})
		}
	}
	sort.Slice(stamps, func(i, j int) bool { return stamps[i].stamp > stamps[j].stamp }) // newest first
	removed := []string{}
	for _, s := range stamps[min(keep, len(stamps)):] {
		if err := os.Remove(filepath.Join(dir, s.name)); err == nil {
			removed = append(removed, s.name)
		}
	}
	return removed
}

// BoardArchivedAlert mirrors src/resilience/operator-alert.ts
// BoardArchivedAlert (Q16 board-archived payload).
//
// TODO(FR-GO-05 #190): replace with the shared internal/resilience
// operator-alert port when it lands (not in this sub-wave's ownership list).
type BoardArchivedAlert struct {
	Event  string `json:"event"`
	Ts     string `json:"ts"`
	Repo   string `json:"repo"`
	Prefix string `json:"prefix"`
	Path   string `json:"path"`
	Reason string `json:"reason"`
}

// OperatorNotifier mirrors src/resilience/operator-alert.ts OperatorNotifier:
// injection seam for outbound paging transports.
type OperatorNotifier func(url string, alert BoardArchivedAlert) error

// postOperatorAlert is the default transport: a best-effort JSON POST.
func postOperatorAlert(url string, alert BoardArchivedAlert) error {
	body, err := json.Marshal(alert)
	if err != nil {
		return err
	}
	resp, err := http.Post(url, "application/json", strings.NewReader(string(body)))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return nil
}

// ArchiveBoardOptions mirrors the TS ArchiveBoardOptions: archiveBoard's
// best-effort Q16 side effects (tests inject them).
type ArchiveBoardOptions struct {
	// URL: operator webhook URL; empty = no paging (opt-in, like Q41).
	URL string
	// Notify: injection seam for tests; defaults to postOperatorAlert.
	Notify OperatorNotifier
	// Now: injectable clock for the alert ts (tests); zero = time.Now.
	Now time.Time
	// Reason: human-readable why, carried onto the board-archived alert;
	// empty = "<prefix> board archived".
	Reason string
	// Keep: retention bound; nil = ArchiveRetentionKeep.
	Keep *int
}

// ArchiveBoard mirrors archiveBoard: move the board into .devagent/archive/
// under `prefix-<stamp>.json`, then fire the best-effort `board-archived`
// operator alert and prune the archive to the retention bound (Q16). Returns
// the repo-relative path. Paging and retention never fail the archive; a
// failed rename does (the CLI maps it to exit 2).
func ArchiveBoard(repoPath, boardPath, prefix, stamp string, opts ArchiveBoardOptions) (string, error) {
	dir := filepath.Join(repoPath, ".devagent", "archive")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	rel := filepath.Join(".devagent", "archive", fmt.Sprintf("%s-%s.json", prefix, stamp))
	if err := os.Rename(boardPath, filepath.Join(repoPath, rel)); err != nil {
		return "", err
	}
	if opts.URL != "" {
		notify := opts.Notify
		if notify == nil {
			notify = postOperatorAlert
		}
		now := opts.Now
		if now.IsZero() {
			now = time.Now()
		}
		reason := opts.Reason
		if reason == "" {
			reason = fmt.Sprintf("%s board archived", prefix)
		}
		// paging is observability, never a failure signal for the loop (Q16)
		_ = notify(opts.URL, BoardArchivedAlert{
			Event:  "board-archived",
			Ts:     now.UTC().Format("2006-01-02T15:04:05.000Z07:00"),
			Repo:   repoPath,
			Prefix: prefix,
			Path:   rel,
			Reason: reason,
		})
	}
	keep := ArchiveRetentionKeep
	if opts.Keep != nil {
		keep = *opts.Keep
	}
	PruneArchive(repoPath, keep)
	return rel, nil
}

// pruneMergedWorktrees mirrors the TS safe-gated merged-worktree prune (the
// shell's cleanup_merged_worktrees): only an executable repo script, never
// fatal.
func pruneMergedWorktrees(repoPath string) {
	script := filepath.Join(repoPath, "scripts", "git-cleanup-merged.sh")
	info, err := os.Stat(script)
	if err != nil || info.IsDir() || info.Mode()&0o111 == 0 {
		return // missing or not executable: nothing to do
	}
	// cleanup is opportunistic; a failed prune must not break the cycle
	_ = exec.Command(script, "--root", repoPath, "--apply").Run()
}

// RunBoardRecoveryOptions mirrors the TS RunBoardRecoveryOptions.
type RunBoardRecoveryOptions struct {
	RecoveryOptions
	// Now: injectable clock for the archive stamp (tests); zero = time.Now.
	Now time.Time
	// Notify: injection seam for tests: outbound operator paging transport (Q16).
	Notify OperatorNotifier
}

// RunBoardRecovery mirrors runBoardRecovery: one recovery cycle — read the
// board, decide, perform the action, return the verdict. Mutates the board
// file (requeue) and moves it (archive) exactly where the shell did. Returns
// an error only when the board vanishes between the decision and the action
// — the CLI maps that to exit 2 (unresolved → the shell falls back to wait).
func RunBoardRecovery(repoPath string, opts RunBoardRecoveryOptions) (BoardRecoveryVerdict, error) {
	boardPath := filepath.Join(repoPath, BoardFile)
	board := readBoard(boardPath)
	if board == nil {
		return BoardRecoveryVerdict{Action: RecoveryActionWait, Reason: "board unreadable"}, nil
	}
	now := opts.Now
	if now.IsZero() {
		now = time.Now()
	}
	stamp := FormatTimestamp(now)
	sleep := fmt.Sprintf("; sleeping %ds", opts.PollSecs)
	// Q16: resolve the paging webhook + retention bound once. A broken config
	// must not turn an archive into a cycle failure, so paging stays opt-in
	// and the retention bound falls back to ArchiveRetentionKeep.
	var url string
	var keep *int
	if cfg, err := config.Load(repoPath); err == nil && cfg.Resilience != nil {
		url = cfg.Resilience.DegradeWebhookURL
		if cfg.Resilience.ArchiveKeep != nil {
			k := int(*cfg.Resilience.ArchiveKeep)
			keep = &k
		}
	}

	counts := CountBoard(boardLikes(board.Tasks))
	intent := DecideBoardRecovery(counts, opts.RecoveryOptions)
	if intent.Kind == IntentWait {
		return BoardRecoveryVerdict{Action: RecoveryActionWait, Reason: intent.Reason}, nil
	}
	if intent.Kind == IntentArchiveComplete {
		rel, err := ArchiveBoard(repoPath, boardPath, "board", stamp, ArchiveBoardOptions{
			URL:    url,
			Keep:   keep,
			Notify: opts.Notify,
			Now:    now,
			Reason: fmt.Sprintf("board complete (%d done)", counts.Done),
		})
		if err != nil {
			return BoardRecoveryVerdict{}, err
		}
		pruneMergedWorktrees(repoPath)
		return BoardRecoveryVerdict{Action: RecoveryActionWait, Reason: fmt.Sprintf("board complete (%d done); archived to %s", counts.Done, rel)}, nil
	}

	rq := RequeueParked(boardPath, board, opts.MaxTotalAttempts)
	post := DecidePostRequeue(CountBoard(boardLikes(board.Tasks)))
	if post.Kind == PostRequeueIntentKindRequeue {
		return BoardRecoveryVerdict{Action: RecoveryActionRequeue, Reason: fmt.Sprintf("reset %d parked task(s) to pending%s", rq.Reset, sleep)}, nil
	}
	// capped > 0 guarantees stuck > 0 here, so refusals always land on this
	// archive verdict: the tasks stay terminal and the loop re-bridges.
	capNote := ""
	if rq.Capped > 0 {
		capNote = fmt.Sprintf("%d task(s) over cumulative attempt cap %d; ", rq.Capped, opts.MaxTotalAttempts)
	}
	rel, err := ArchiveBoard(repoPath, boardPath, "board-stuck", stamp, ArchiveBoardOptions{
		URL:    url,
		Keep:   keep,
		Notify: opts.Notify,
		Now:    now,
		Reason: capNote + post.Detail,
	})
	if err != nil {
		return BoardRecoveryVerdict{}, err
	}
	return BoardRecoveryVerdict{Action: RecoveryActionArchive, Reason: fmt.Sprintf("reset %d parked task(s) to pending; %s%s; archived to %s", rq.Reset, capNote, post.Detail, rel)}, nil
}

// boardLikes adapts raw board task maps to the loose BoardTaskLike shape.
func boardLikes(tasks []map[string]any) []BoardTaskLike {
	likes := make([]BoardTaskLike, len(tasks))
	for i, t := range tasks {
		if t == nil {
			continue
		}
		likes[i] = BoardTaskLike{Status: t["status"], Attempts: t["attempts"], TotalAttempts: t["totalAttempts"]}
	}
	return likes
}

// looseString mirrors the TS String(x) coercion used for loose status checks.
func looseString(v any) string {
	switch x := v.(type) {
	case nil:
		return "undefined"
	case string:
		return x
	default:
		return fmt.Sprintf("%v", x)
	}
}

// numberIsFinite mirrors Number(x) + Number.isFinite for the totalAttempts
// check. (Distinctly named from autopr.go's jsNumber, which returns the bare
// Number(x) value without the finite probe.)
func numberIsFinite(v any) (float64, bool) {
	switch x := v.(type) {
	case nil:
		return 0, false // Number(undefined) is NaN
	case float64:
		return x, true
	case int:
		return float64(x), true
	case bool:
		if x {
			return 1, true
		}
		return 0, true
	case string:
		s := strings.TrimSpace(x)
		if s == "" {
			return 0, true // Number('') is 0
		}
		f, err := strconv.ParseFloat(s, 64)
		if err != nil {
			return 0, false
		}
		return f, true
	default:
		return 0, false
	}
}

// containsString is a tiny membership check.
func containsString(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}
